File: aat-android/src/main/java/ch/bailu/aat/dispatcher/CurrentLocationSource.java
package ch.bailu.aat.dispatcher;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat.util.OldAppBroadcaster;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.dispatcher.ContentSource;
import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.gpx.InfoID;

public class CurrentLocationSource extends ContentSource {
    private final ServiceContext scontext;

    private final BroadcastReceiver onLocationChange = new BroadcastReceiver () {
        @Override
        public void onReceive(Context context, Intent intent) {
            sendUpdate(InfoID.LOCATION, scontext.getLocationService().getLocationInformation());
        }

    };



    public CurrentLocationSource(ServiceContext sc) {
        scontext=sc;
    }





    @Override
    public void requestUpdate() {
        sendUpdate(InfoID.LOCATION, scontext.getLocationService().getLocationInformation());
    }


    @Override
    public void onPause() {
        scontext.getContext().unregisterReceiver(onLocationChange);
    }

    @Override
    public void onResume() {
        OldAppBroadcaster.register(scontext.getContext(), onLocationChange,
                AppBroadcaster.LOCATION_CHANGED);
    }

    @Override
    public int getIID() {
        return InfoID.LOCATION;
    }

    @Override
    public GpxInformation getInfo() {
        return scontext.getLocationService().getLocationInformation();
    }

}

File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/source/CurrentLocationSource.kt
package ch.bailu.aat_lib.dispatcher.source

import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.broadcaster.Broadcaster
import ch.bailu.aat_lib.dispatcher.SourceInterface
import ch.bailu.aat_lib.dispatcher.TargetInterface
import ch.bailu.aat_lib.gpx.GpxInformation
import ch.bailu.aat_lib.gpx.InfoID
import ch.bailu.aat_lib.service.ServicesInterface

class CurrentLocationSource(
    private val services: ServicesInterface,
    private val broadcaster: Broadcaster
) : SourceInterface {

    private var target = TargetInterface.NULL
    private val onLocationChange = BroadcastReceiver { requestUpdate() }

    override fun setTarget(target: TargetInterface) {
        this.target = target
    }

    override fun requestUpdate() {
        target.onContentUpdated(InfoID.LOCATION, services.getLocationService().getLocationInformation())
    }

    override fun onPauseWithService() {
        broadcaster.unregister(onLocationChange)
    }

    override fun onDestroy() {}

    override fun onResumeWithService() {
        broadcaster.register(AppBroadcaster.LOCATION_CHANGED, onLocationChange)
    }

    override fun getIID(): Int {
        return InfoID.LOCATION
    }

    override fun getInfo(): GpxInformation {
        return services.getLocationService().getLocationInformation()
    }
}
